src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -633,17 +724,56 @@ private static Expression BindParameterFromRouteValueOrQueryString(ParameterInfo
             return BindParameterFromValue(parameter, Expression.Coalesce(routeValue, queryValue), factoryContext);
         }
 
-        private static Expression BindParameterFromBody(Type parameterType, bool allowEmpty, FactoryContext factoryContext)
+        private static Expression BindParameterFromBody(ParameterInfo parameter, bool allowEmpty, FactoryContext factoryContext)
         {
             if (factoryContext.JsonRequestBodyType is not null)
             {
                 throw new InvalidOperationException("Action cannot have more than one FromBody attribute.");
             }
 
-            factoryContext.JsonRequestBodyType = parameterType;
-            factoryContext.AllowEmptyRequestBody = allowEmpty;
+            var nullability = nullabilityContext.Create(parameter);
+            var isOptional = parameter.HasDefaultValue || nullability.ReadState == NullabilityState.Nullable;
+
+            factoryContext.JsonRequestBodyType = parameter.ParameterType;
+            factoryContext.AllowEmptyRequestBody = allowEmpty || isOptional;
+
+            var argument = Expression.Variable(parameter.ParameterType, $"{parameter.Name}_local");
+
+            if (!factoryContext.AllowEmptyRequestBody)
+            {
+                // If the parameter is required or the user has not explicitly
+                // set allowBody to be empty then validate that it is required.
+                //
+                // ToDo body_local = Convert(bodyValue, ToDo);
+                // if (body_local == null)
+                // {
+                //      wasTryParseFailure = true;
+                //      Log.RequiredParameterNotProvided(httpContext, "Todo", "body")
+                // }
+                var checkRequiredBodyBlock = Expression.Block(
+                    Expression.Assign(argument, Expression.Convert(BodyValueExpr, parameter.ParameterType)),
+                    Expression.IfThen(Expression.Equal(argument, Expression.Constant(null)),
+                        Expression.Block(
+                            Expression.Assign(WasTryParseFailureExpr, Expression.Constant(true)),
+                            Expression.Call(LogRequiredParameterNotProvidedMethod, 
+                                    HttpContextExpr, Expression.Constant(parameter.ParameterType.Name), Expression.Constant(parameter.Name))
+                        )
+                    )
+                );
+                factoryContext.TryParseParams.Add((argument, checkRequiredBodyBlock));
```

Comment:
I think this should also be replaced by a `bool FactoryContext.CheckBindingFailures`.

Reply:
Nvm. As discussed offline, this wouldn't work because we need also to add these checking expressions earlier in the method body.